common/base/include/claragenomics/utils/threadsafe_containers.hpp

```diff
@@ -0,0 +1,222 @@
+    /// \throw std::logic_error if called after signal_pushed_last_element() has been called by any producer
+    void add_new_element(const T& element) // not using reference in order to support both rvalue and lvalue
+    {
+        // consider adding max number of elements and waiting for consumer to process an element before adding a new one
+        std::lock_guard<std::mutex> lg(mutex_);
+        if (pushed_last_element_)
+        {
+            throw std::logic_error("ThreadsafeProducerConsumer: pushed an element after signal_pushed_last_element() has been called");
+        }
+        data_.push_front(element);
+        condition_variable_.notify_all();
+    }
+
+    /// \brief adds an element to the queue
+    ///
+    /// \param element element to add
+    /// \throw std::logic_error if called after signal_pushed_last_element() has been called by any producer
+    void add_new_element(T&& element) // not using reference in order to support both rvalue and lvalue
+    {
+        // consider adding max number of elements and waiting for consumer to process an element before adding a new one
+        std::lock_guard<std::mutex> lg(mutex_);
+        if (pushed_last_element_)
+        {
+            throw std::logic_error("ThreadsafeProducerConsumer: pushed an element after signal_pushed_last_element() has been called");
+        }
+        data_.push_front(std::move(element));
+        condition_variable_.notify_all();
+    }
+
+    /// \brief tells container that no new elements are going to be added
+    ///
+    /// After this method has been called and all elements have been consumed get_next_element() will be returning
+    /// empty optionals to indicate that there will be no new elements and that consumers can finish processing
+    ///
+    /// \throw std::logic_error if called after any producer has already called this method
+    void signal_pushed_last_element()
+    {
+        std::lock_guard<std::mutex> lg(mutex_);
+        if (pushed_last_element_)
+        {
+            throw std::logic_error("ThreadsafeProducerConsumer: called signal_pushed_last_element() more than once");
+        }
+        pushed_last_element_ = true;
+        condition_variable_.notify_all();
+    }
+
+    /// \brief gets the next element
+    ///
+    /// If no element is available waits for an element to become available
+    /// If no element is available and signal_pushed_last_element() has already been called returns an empty optional
+    ///
+    /// \return optional, either with the value or empty if no element available and signal_pushed_last_element() has been called
+    cga_optional_t<T> get_next_element()
+    {
+        std::unique_lock<std::mutex> ul(mutex_);
+
+        while (data_.empty() && !pushed_last_element_)
+        {
+            condition_variable_.wait(ul);
+        }
+
+        const bool no_elements_left = pushed_last_element_ && data_.empty();
+        if (no_elements_left)
+        {
+#ifdef __cpp_lib_optional
```

Comment:
Compiler complains about this, which makes sense, wouldn't this be exactly like saying `return int;`?

Reply:
Oh yes, good point. I think it would make sense to define a global `constexpr cga_nullopt = std[::experimental]::nullopt;` together with the type then. What's your opinion on that?

Reply:
Yes, `constexpr` worked. I'll submit it now and see if it also works with GCC 5.4

Reply:
Great, thanks @mimaric !